src/components/tl/ucp/alltoallv/alltoallv_hybrid.c

```diff
@@ -0,0 +1,1065 @@
+    uint32_t           radix  = task->alltoallv_hybrid.radix;
+    ucc_rank_t         tsize  = UCC_TL_TEAM_SIZE(team);
+    ucc_tl_ucp_alltoallv_hybrid_buf_meta_t *lbm;
+    int i;
+
+    lbm = TASK_LB(task, tsize);
+    UCC_TL_UCP_PROFILE_REQUEST_EVENT(coll_task, "ucp_alltoallv_hybrid_start",
+                                     0);
+    ucc_tl_ucp_task_reset(task, UCC_INPROGRESS);
+
+    task->alltoallv_hybrid.num_in      = 0;
+    task->alltoallv_hybrid.cur_out     = 1;
+    task->alltoallv_hybrid.cur_radix   = 1;
+    task->alltoallv_hybrid.traffic_in  = 0;
+    task->alltoallv_hybrid.traffic_out = 0;
+    task->alltoallv_hybrid.num2send    = 0;
+    task->alltoallv_hybrid.num2recv    = 0;
+    task->alltoallv_hybrid.phase       = UCC_ALLTOALLV_HYBRID_PHASE_START;
+    task->alltoallv_hybrid.iteration   = 1;
+
+    memset(TASK_SEG(task, tsize), 0, tsize*sizeof(char));
+    for (i = 0; i < radix - 1; ++i){
+        meta_init(&lbm[i], task);
+    }
+    return ucc_progress_queue_enqueue(UCC_TL_CORE_CTX(team)->pq, &task->super);
+}
+
+
+static inline void copy_brucks_rotation(void *scratch,
+                                        void *scounts, void *sdisps,
+                                        ucc_rank_t trank, ucc_rank_t tsize)
+{
+
+    size_t size = sizeof(int); /* currently support for 32 bit counts */
+
+    memcpy(scratch, PTR_OFFSET(sdisps, trank * size), (tsize - trank) * size);
+    memcpy(PTR_OFFSET(scratch, tsize * size), PTR_OFFSET(scounts, trank * size),
+           (tsize - trank) * size);
+    /* Copy the rest part */
+    if (trank != 0) {
+        memcpy(PTR_OFFSET(scratch, (tsize - trank) * size), sdisps, trank * size);
+        memcpy(PTR_OFFSET(scratch, (tsize - trank + tsize) * size), scounts,
+               trank * size);
+    }
+}
+
+
+ucc_status_t ucc_tl_ucp_alltoallv_hybrid_init(ucc_base_coll_args_t *coll_args,
+                                              ucc_base_team_t      *team,
+                                              ucc_coll_task_t     **task_h)
+{
+    ucc_tl_ucp_team_t *tl_team    = ucc_derived_of(team, ucc_tl_ucp_team_t);
+    ucc_rank_t         tsize      = UCC_TL_TEAM_SIZE(tl_team);
+    ucc_rank_t         trank      = UCC_TL_TEAM_RANK(tl_team);
+    uint32_t           radix      = UCC_TL_UCP_TEAM_LIB(tl_team)->cfg.alltoallv_hybrid_radix;
+    size_t             buff_size  = UCC_TL_UCP_TEAM_LIB(tl_team)->cfg.alltoallv_hybrid_buff_size;
+    uint32_t           snd_size   = UCC_TL_UCP_TEAM_LIB(tl_team)->cfg.alltoallv_hybrid_send_buffer_size;
+    uint32_t           rcv_size   = UCC_TL_UCP_TEAM_LIB(tl_team)->cfg.alltoallv_hybrid_recv_buffer_size;
+    ucc_tl_ucp_task_t *task;
+    size_t             scratch_size, calc_limit, max_snd_count, dt_size;
+    ucc_status_t       status;
+
+    if (UCC_COLL_ARGS_DISPL64(&coll_args->args) ||
+        UCC_COLL_ARGS_COUNT64(&coll_args->args) ||
+        coll_args->args.src.info_v.mem_type != UCC_MEMORY_TYPE_HOST ||
```

Comment:
just to kick off the discussion: what would it take to enable this for GPU mem? lets say we will still use host scratch. so first non-pairwise send potentially goes GPU->network->host via TL/UCP. Then we only need to fix buffer_recycler, wait_for_recv : memcpy -> copy_host_2_device. Could this give advantage over default GPU pairwise ?

2nd discussion: how (do we need to) support 64bit counts? One obvious solution is to make the first tsize * 2 entries of scratch to always be uint64_t. It would double the mem consumption, but is it critical? Currently, single mpool entry is 1Mb by default (iirc). Then, we have radix-1 (radix=2) of hybrid_buff_size = 256K for actual scratch. This is just 256K. we have 768K left for all the rest. Say, team size is 32K (quite large job, but not super scale obviously). then 32K * 2 * int = 256K, while same for uin64 would be 512 K. So it will still fit into same scratch mpool chunk. so this space is essentially free anyways. 65K team will not fit, will require larger scratch.

Ideal solution is of course jus tallocate space based on 64bit flag stored on task. And replace all the direct acceses to the inline getter/setter that would go with the proper offset